libgraph/include/katana/ArrowRandomAccessBuilder.h

```diff
@@ -1,76 +1,111 @@
-#include <arrow/api.h>
+#include <arrow/array.h>
+#include <arrow/builder.h>
 
-#include "katana/ErrorCode.h"
 #include "katana/Logging.h"
-#include "katana/NUMAArray.h"
-#include "katana/Properties.h"
 #include "katana/Result.h"
 
 namespace katana {
 
-namespace {
+/// ArrowRandomAccessBuilder encapsulates the concept of building
+/// an arrow::Array from <index, value> pairs arriving in unknown order
+/// Thread-safety guarantee is equivalent to that of std::vector<uint8_t>
+/// Concurrent accesses to different indices are thread safe
+/// Concurrent accesses to a single index are not safe
+///
+/// Example Usage:
+/// katana::ArrowRandomAccessBuilder<arrow::Int64Type> builder{128};
+/// builder[55] = 7;
+/// builder[127] = 11;
+/// std::shared_ptr<arrow::Array> array =
+///     KATANA_CHECKED(builder.Finalize());
+/// array has 128 elements, 2 are valid, the rest are null
+///
+/// API:
+/// template <typename ArrowType>
+/// class ArrowRandomAccessBuilder {
+///   value_type& operator[](index);
+///   void UnsetValue(index);
+///   bool IsValid(index);
+///   katana::Result<std::shared_ptr<arrow::Array>> Finalize();
+/// };
+///
+/// TODO(danielm) express this API through inheritance or concepts
+
+namespace internal {
 
-/// NoNullBuilder uses std::vector for storage
-/// Finalize() makes a copy of the data
-/// Does not support null values
-template <typename ValueType, typename StorageType, typename ArrowType>
-class NoNullBuilder {
+template <typename ArrowType>
+class FixedSizeBuilder {
 public:
-  using value_type = ValueType;
-  using reference = ValueType&;
+  using value_type = typename ArrowType::c_type;
 
-  NoNullBuilder(size_t length) : data_(length) {
-    static_assert(sizeof(ValueType) == sizeof(StorageType));
+  katana::Result<void> Resize(size_t length) {
+    KATANA_CHECKED(data_.Resize(length));
+    data_.bytes_builder()->UnsafeAdvance(
+        data_.bytes_builder()->capacity() - data_.bytes_builder()->length());
+    valid_.resize(length, false);
+    return katana::ResultSuccess();
   }
 
-  reference operator[](size_t index) {
+  // NOTE this operator has side-effects. It can safely be used in two ways:
+  // 1) builder[index] = value; where it creates a non-null entry
+  // 2) value = builder[index]; ONLY IF option 1 has already used that index
+  value_type& operator[](size_t index) {
     KATANA_LOG_DEBUG_VASSERT(
         index < size(), "index: {}, size: {}", index, size());
-    return static_cast<ValueType*>(data_.data())[index];
+    valid_[index] = true;
+    return reinterpret_cast<value_type*>(data_.mutable_data())[index];
```

Comment:
Is the operator used for setting or getting data, or both?
Is a check for validity needed before getting the data?

Reply:
This is mostly used for setting data, but it can be used to read data from already-set positions. Once we return a reference to the caller we can't really control which way it's used without building a bunch of wrapper logic that would add overhead